File: 0011-container-with-most-water/my_approach.cpp
// Time complexity: O(n), Space complexity: O(1)
// The objective is to maximize area and hence maximize height and maximize width
// Starting at the maximum width, l = 0, r = size - 1, we move towards maximum height, as it is
// bound by min(a[l], a[r]). Keep moving pointer pointing to lower height.
class Solution {
public:
    int maxArea(vector<int>& height) {
        int size = height.size();
        int l = 0, r = size - 1, h, w, currArea = -1;
        int maxArea = -1;

        while(l < r && r >= 0) {
            h = min(height[l], height[r]);
            w = r - l;
            currArea = h*w;
            if (height[l] < height[r]) {
                l++;
            }
            else {
                r--;
            }
            maxArea = max(maxArea, currArea);
        }
        return maxArea;
    }
};

File: 0011-container-with-most-water/my_approach.py
# Time complexity: O(n), Space complexity: O(1)
class Solution:
    def maxArea(self, height: List[int]) -> int:
        size = len(height)
        max_area = curr_area = -1
        l, r = 0, size - 1

        while l < r:
            h = min(height[l], height[r])
            w = r - l
            curr_area = h*w
            max_area = max(max_area, curr_area)
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        
        return max_area

            